### ISSUE TITLE
self.all_gather collects more data than it acutally is

### ISSUE THREAD

**Issue by zhiyuanpeng** [PyTorchLightning/pytorch-lightning#11102]
## 🐛 Bug

I have 3250 dev instances. In single GPU mode, in function `validation_epoch_end`, after collecting all the resutls from `outputs`, I get 3250 instances. However, in `ddp`mode, I train my model on 4 gpus and after `self.all_gather(outputs)` in `validation_epoch_end`, I get 3252 instances. 
### To Reproduce

```
    def validation_epoch_end(self, outputs: List[Any]):
        if self.trainer.gpus == -1:
            outputs = self.all_gather(outputs)
        y_true, y_pred = [], []
        rows_acc = 0
        for out_dict in outputs:
            if self.trainer.gpus == -1:
                gpu_count, batch_size, max_len = out_dict["sents_label_id"].size()
                sents_label_id = out_dict["sents_label_id"].reshape(
                    gpu_count * batch_size, max_len
                )
                preds = out_dict["preds"].reshape(gpu_count * batch_size, max_len)
                sents_label_mask = out_dict["sents_label_mask"].reshape(
                    gpu_count * batch_size, max_len
                )
                sents_frist_subtoken_loc = out_dict["sents_frist_subtoken_loc"].reshape(
                    gpu_count * batch_size, max_len
                )
                rows_acc += gpu_count*batch_size
            else:
                sents_label_id = out_dict["sents_label_id"]
                preds = out_dict["preds"]
                sents_label_mask = out_dict["sents_label_mask"]
                sents_frist_subtoken_loc = out_dict["sents_frist_subtoken_loc"]
            p_y_true, p_y_pred = ids_to_labels(
                sents_label_id.tolist(),
                preds.tolist(),
                self.label_list,
                sents_label_mask.tolist(),
                sents_frist_subtoken_loc,
                self.hparams.feature_mode,
            )
            y_true.extend(p_y_true)
            y_pred.extend(p_y_pred)
        assert rows_acc == 3250
```

### Expected behavior
No matter in single GPU mode or `ddp` mode, `len(y_true)`==`len(y_pred)`==3250

### Environment

* CUDA:
        - GPU:
                - Tesla V100-SXM2-16GB
                - Tesla V100-SXM2-16GB
                - Tesla V100-SXM2-16GB
                - Tesla V100-SXM2-16GB
        - available:         True
        - version:           10.2
* Packages:
        - numpy:             1.19.5
        - pyTorch_debug:     False
        - pyTorch_version:   1.10.0+cu102
        - pytorch-lightning: 1.5.5
        - tqdm:              4.62.3
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.8.10
        - version:           #135-Ubuntu SMP Fri Jan 19 11:48:36 UTC 2018

**Comment by rohitgr7** [comment]
is drop_last=False in your dataloader?
possibly because DistributedSampler behind the scene adds some extra samples to make them evenly distributed across devices.
https://github.com/pytorch/pytorch/blob/73a6c36f1bfbf9aff04ba41cfe6ab06aa99883d9/torch/utils/data/distributed.py#L103-L109

**Comment by zhiyuanpeng** [comment]
> is drop_last=False in your dataloader? possibly because DistributedSampler behind the scene adds some extra samples to make them evenly distributed across devices. https://github.com/pytorch/pytorch/blob/73a6c36f1bfbf9aff04ba41cfe6ab06aa99883d9/torch/utils/data/distributed.py#L103-L109

Thanks, yes, I set drip_last=False, because it's a research project, I don't want to loss any data. Any suggestions to solve this problem: both DDP and not loss any data? Thanks.

**Comment by zhiyuanpeng** [comment]
> is drop_last=False in your dataloader? possibly because DistributedSampler behind the scene adds some extra samples to make them evenly distributed across devices. https://github.com/pytorch/pytorch/blob/73a6c36f1bfbf9aff04ba41cfe6ab06aa99883d9/torch/utils/data/distributed.py#L103-L109

Thanks, I have found the solution https://github.com/PyTorchLightning/pytorch-lightning/issues/8375
